src/EventLoop.cpp

```diff
@@ -155,9 +165,18 @@ auto EventLoop::Stop() -> void
     m_is_stopping = true;
 }
 
-auto EventLoop::GetActiveRequestCount() const -> uint64_t
+static auto has_pending_requests(std::mutex& mutex, const std::vector<RequestHandle>& pending_requests) -> bool
+{
+    std::unique_lock lock{mutex};
+    return !pending_requests.empty();
+}
+
+auto EventLoop::HasUnfinishedRequests() -> bool
 {
-    return m_active_request_count;
+    return (
+           (m_active_request_count.load() > 0)
```

Comment:
thinking about this there's still a race condition. you can load the value before you get the mutex (so the loop thread could have the lock and be in the middle of moving pending->active). Then you block waiting for the loop thread to release the lock, and when they do you get an empty pending requests, but you didn't count those requests as active either. I think the solution is probably unfortunately just holding the lock for this entire function, not just for the pending requests check. But you'd have to check the loop logic to see if that's safe either

Reply:
Good point -- looks like we need to acquire the lock at the beginning of the function